### PR TITLE
[refactor] Use package_layout instead of passing directories (source command and binary_installer)

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: omit
Docs: omit

This is a draft, there are a couple of things I don't like too much:
 * `package_layout.build(pref)` returns the build directory for the `pref.id`, but it could be different if the recipe has defined `build_id()` method... Once again, the layout depends on the sources of the `conanfile.py` itself, I wonder if we should provide the working `conanfile` object to the layout.
 * I've created a `PackageLocalLayout` which is handy for the local workflow, but... meh

### CODE REVIEW

File: conans/util/files.py [hunk 1]
```diff
@@ -334,3 +334,38 @@ def exception_message_safe(exc):
...
+    for src_dir, dirs, files in walk(src, followlinks=True):
+        if is_excluded(src_dir):
+            dirs[:] = []
+            continue
+
+        # Overwriting the dirs will prevents walk to get into them
+        files[:] = [d for d in files if not is_excluded(os.path.join(src_dir, d))]
+
+        dst_dir = os.path.normpath(os.path.join(dst, os.path.relpath(src_dir, src)))
+        if not os.path.exists(dst_dir):
```
Comment (SSE4) [inline conans/util/files.py]: `os.path.isdir`? also, what if it exists, but as file?

File: conans/client/source.py [hunk 2]
```diff
@@ -9,7 +9,21 @@
...
+                               merge_directories)
+
+
+class PackageLocalLayout(object):
+    ref = None
+
+    def __init__(self, base_folder):
+        self._base_folder = base_folder
+
+    def base_folder(self):
```
Comment (SSE4) [inline conans/client/source.py]: property?

File: conans/client/source.py [hunk 3]
```diff
@@ -9,7 +9,21 @@
...
+        return self._base_folder
+
+    def conanfile(self):
```
Comment (SSE4) [inline conans/client/source.py]: property?

File: conans/client/source.py [hunk 4]
```diff
@@ -9,7 +9,21 @@
 from conans.model.conan_file import get_env_context_manager
 from conans.model.scm import SCM, get_scm_data
 from conans.paths import CONANFILE, CONAN_MANIFEST, EXPORT_SOURCES_TGZ_NAME, EXPORT_TGZ_NAME
-from conans.util.files import (clean_dirty, is_dirty, load, mkdir, rmdir, set_dirty, walk)
+from conans.util.files import (clean_dirty, is_dirty, load, mkdir, rmdir, set_dirty,
+                               merge_directories)
+
+
+class PackageLocalLayout(object):
+    ref = None
```
Comment (SSE4) [inline conans/client/source.py]: seems like not used anywhere?

Comment (jgsogo) [reply conans/client/source.py]: Yes, it is used [here](https://github.com/conan-io/conan/pull/4901/files#diff-802aae676800bf9b0e36c4909618951dR117)

File: conans/client/source.py (hunk 2 above)
Comment (jgsogo) [reply conans/client/source.py]: Have a look at `PackageCacheLayout`, this class belongs to the same family

File: conans/client/source.py (hunk 3 above)
Comment (jgsogo) [reply conans/client/source.py]: Same as above, this is a function in `PackageCacheLayout`

File: conans/util/files.py (hunk 1 above)
Comment (jgsogo) [reply conans/util/files.py]: Agree with you, it deserves another PR, in this one I don't want to touch anything related to functionality. Also this PR is just a _draft_ and probably it won't get merged, I'm thinking about dividing it into two PR (this one has changes affecting the `source` command and other targeting the `binary_installer`, I prefer not to mix)

**Comment by jgsogo** [comment]
Not for now, [moved to my workspace](https://github.com/jgsogo/conan/pull/12)

**Pull request closed without merging**